man/vipsprofile.1
.TH VIPSPROFILE 1 "13 December 2013"
.SH NAME
vipsprofile \- analyze vips profiles
.SH SYNOPSIS
.B vipsprofile
.SH DESCRIPTION
.B vipsprofile(1)
analyzes the file written by the --vips-profile option, calculates some
statistics, and draws a graph of evaluation.

Run any vips program with the --vips-profile option to generate a file called
"vips-profile.txt". This contains timing information about CPU use, memory use
and thread synchronisation. 

Run
.B vipsprofile(1)
to load this file, calculate some 
statistics, and draw a graph of evaluation saved to vips-profile.svg. This
analysis can help track down performance problems. 

For example:

  $ vips sharpen shark.jpg x.jpg --vips-profile
  recording profile in vips-profile.txt
  $ vipsprofile
  reading from vips-profile.txt
  loaded 3622 events
  total time = 0.138322
  name          alive   wait%   work%   unkn%   memory  peakm
  worker 20      0.069  34.5    58.9    6.65    3.14    5.56    
  worker 21       0.07  1.36    60.2    38.4    2.65    5.07    
  worker 22       0.07  33      55.8    11.1    2.62    5.04    
  worker 23      0.072  34.2    59.7    6.15    2.72    5.14    
  wbuffer 24     0.075  99      1.03    0.00401 0       0       
  wbuffer 25     0.075  95.6    4.39    0.00667 0       0       
  main 26         0.14  52.8    0       47.2    -11.1   0.787   
  peak memory = 21.6 MB
  writing to vips-profile.svg

.SH RETURN VALUE
returns 0 on success and non-zero on error.
.SH SEE ALSO
vips(1)


NEWS
VIPS changed from 7.10 to 7.12 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- IMAGE->Hist is deprecated, use im_history_get() instead
- im_append_Hist() removed
- IMAGE->Bbits is deprecated (now ignored)
- im_region_local() replaced by im_region_buffer()

VIPS enhancements
=================

- new backwards and forwards compatible VIPS file format: it has a new 
  metadata system that efficiently shares and propogates ICC profiles,
  EXIF data, etc. & whatever
- speed ups: new pixel buffer cache saves and reuses computations, uses liboil 
  where possible for a small speedup
- SMP scaling: simplified lock system improves SMP scaling, double-buffered 
  image writes overlap computation 
- lower mem use: new mmap window cache shares file mappings, history buffers 
  share and reuse image history metadata
- built-in support for new image formats: OpenEXR, Analyze7, CSV
- RGB16 and GREY16 image hints
- uses GOption for much better command-line argument handling
- better C++ binding: namespaces, VError derives from std::exception, image
  and number vector constants
- python binding 
- gcc sentinel attributes added
- added GValue as an arg type
- added im_benchmark() and friends for testing
- new functions: 
	im_analyze2vips(), im_benchmark*(), im_buildlut(),
	im_concurrency_get/set(), im_contrast_surface(),
	im_contrast_surface_raw(), im_copy_from(), im_copy_morph(),
	im_cp_desc_array(), im_cp_descv(), im_extract_areabands(),
	im_flood_blob_copy(), im_get_option_group(), im_grid(),
	im_header_exists(), im_header_map(), im_history_get(),
	im_invalidate(), im_isscalar(), im_lineset(), im_linreg(),
	im_meta*(), im_msb(), im_norm_dmask(), im_project(),
	im_raw2vips(), IM_REGION_ADDR_TOPLEFT(), im_tile_cache(),
	im_video_test()


VIPS changed from 7.8 to 7.10 (not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- no longer generates vips-config script, instead it makes .pc files for
  pkg-config ... replace any "vips-config" with "pkg-config vips-7.10"
- origin hint semantics changed ... it now records the position in the output
  image of the origin of the input image
- all float to int casts on pixel values are now floor() rather than rint(),
  for consistency with C casting semantics

VIPS enhancements
=================

- nip reworked for gtk+-2.4, now called nip2
- now based on glib, so VIPS no longer has it's own thread, plugin and data
  structure libraries
- i18n support, although not quite complete yet
- new error message API to make i18n possible (compatibility macros mean the
  old API still works)
- new 'start up VIPS' call, im_init_world(), sets up i18n and threads
  (optional to call it, but i18n won't fully work unless you do)
- C programs now just need "#include <vips/vips.h>", no other includes
  required
- wraps the Pango font rendering library, see im_text()
- new function flag: NOCACHE ... means the result of the call should not be
  cached, useful for video grab functions and paintbox functions
- im_embed() now supports a wider range of embedding styles, including a fast
  edge stretch
- all area operations use the new embed style to expand their input rather 
  than their output, so there are no more black borders on filtered images
- other new functions: im_render(), im_cache(), im_extract_bands(), 
  im_copy_swap(), im_rint(), im_make_xy(), im_init_world(), im_error(),
  im_warn(), im_diag(), im_rank_image(), im_replicate()
- many fixes and improvements to old funcs
- configure and portability improvements

Bug fixes
=========

- all serious bug fixes got back-ported to 7.8, I think


VIPS changes from 7.6 to 7.8 (again, not exhaustive, see ChangeLog for details)

Non-backwards compatible changes
================================

- output format options specified in filenames: new prototypes for load and
  save tiff and jpeg
- C++ API name changes ... error => VError
- include path change
    - C includes <vips/vips.h>
    - C++ includes <vips/vipscpp.h>
- im_extract*() numbers bands from zero, not from 1 ... also other funcs which
  refer to bands (eg. im_lrmosaic() etc.)
- mosaic funcs have extra max blend width parameter

VIPS enhancements
=================

- rewritten and much fancier ip (see separate docs), now called nip ... old ip
  (slightly fixed) still there
- mmap windows remove 2GB image size limit
- images have an origin hint field
- icc transforms supported via wrapping of Little CMS
- much, much faster Fourier transforms via wrapping of libfwfft
- ppm/pgm/pbm read/write
- C++ API enhancements: new constructors, more operator equivalences
- new colourspace: Yxy
- support for D50 colour temperature images
- new functions: im_image(), im_system(), im_version*(), im_blend(),
  im_copy_set(), im_flood_blob(), im_icc_*(), im_open_local_array(),
  im_header_*(), im_sign(), im_ceil(), im_floor(), im_remainderconst*(),
  im_print(), im_guess_prefix(), im_remosaic(), im_invertlut(), Yxy funcs,
  color temperature funcs, im_clip2fmt(), im_blend(), im_lab_morph(),
  im_histnorm(), im_histcum(), im_video_v4l(), im_LabS2Lab(), im_Lab2LabS()
- new type: IMAGEVEC
- header is much faster
- ip/nip split to separate packages
- better vips2dj
- better cygwin support, mingw support too
- Mac OS X and Windows build support
- new set of sensible #define names, old names can be turned off with
  -DIM_NO_VIPS7_COMPAT
- many configure improvements

Bug fixes
=========

- speed ups
- fixes to implicit image format conversions 
- im_zoom() spots integer overflow
- im_LabS2LabQ() rounding on a/b slightly broken for a/b == 0
- fixes to refcounting in C++ API
- mask casts in C++ were broken




VIPS Changes for 7.4 to 7.6

Non-backwards compatible changes
================================

- im_histplot() has new (and better) rules
- im_andconst(), im_orconst() and im_eorconst() now take a
  double, not an unsigned char, as their constant argument type

tools/vipsprofile.py
#!/usr/bin/python

import re
import math
import cairo

class ReadFile:
    def __init__(self, filename):
        self.filename = filename

    def __enter__(self):
        self.f = open(self.filename, 'r') 
        self.lineno = 0
        self.getnext();
        return self

    def __exit__(self, type, value, traceback):
        self.f.close()

    def __nonzero__(self):
        return self.line != ""

    def getnext(self):
        self.lineno += 1
        self.line = self.f.readline()

def read_times(rf):
    times = []

    while True:
        match = re.match('[0-9]+ ', rf.line)
        if not match:
            break
        times += [int(x) for x in re.split(' ', rf.line.rstrip())]
        rf.getnext()

    return times[::-1]

class Thread:
    thread_number = 0

    def __init__(self, thread_name):
        self.thread_name = thread_name
        self.thread_number = Thread.thread_number
        self.events = []
        Thread.thread_number += 1

class Event:
    def __init__(self, thread, gate_name, start, stop):
        self.thread = thread
        self.gate_name = gate_name
        self.start = start
        self.stop = stop

        self.work = False
        self.wait = False
        if re.match('.*?: .*work.*', gate_name):
            self.work = True
        if re.match('.*?: .*wait.*', gate_name):
            self.wait = True

        thread.events.append(self)

input_filename = 'vips-profile.txt'

thread_id = 0
threads = []
n_events = 0
print 'reading from', input_filename
with ReadFile(input_filename) as rf:
    while rf:
        if rf.line.rstrip() == "":
            rf.getnext()
            continue
        if rf.line[0] == "#":
            rf.getnext()
            continue

        match = re.match('thread: (.*)', rf.line)
        if not match:
            print 'parse error line %d, expected "thread"' % rf.lineno
        thread_name = match.group(1) + " " + str(thread_id)
        thread_id += 1
        thread = Thread(thread_name)
        threads.append(thread)
        rf.getnext()

        while True:
            match = re.match('gate: (.*)', rf.line)
            if not match:
                break
            gate_name = match.group(1)
            match = re.match('vips_(.*)', gate_name)
            if match:
                gate_name = match.group(1)
            rf.getnext()

            match = re.match('start:', rf.line)
            if not match:
                continue
            rf.getnext()

            start = read_times(rf)

            match = re.match('stop:', rf.line)
            if not match:
                continue
            rf.getnext()

            stop = read_times(rf)

            if len(start) != len(stop):
                print 'start and stop length mismatch'

            for a, b in zip(start, stop):
                Event(thread, gate_name, a, b)
                n_events += 1

for thread in threads:
    thread.events.sort(lambda x, y: cmp(x.start, y.start))

print 'loaded %d events' % n_events

# normalise time axis to secs of computation
ticks_per_sec = 1000000.0
start_time = threads[0].events[0].start
last_time = 0
for thread in threads:
    for event in thread.events:
        event.start = (event.start - start_time) / ticks_per_sec
        event.stop = (event.stop - start_time) / ticks_per_sec

        if event.stop > last_time:
            last_time = event.stop

print 'last time =', last_time

# do two gates overlap?
def is_overlap(events, gate_name1, gate_name2):
    for event1 in events:
        if event1.gate_name != gate_name1:
            continue

        for event2 in events:
            if event2.gate_name != gate_name2:
                continue

            # if either endpoint of 1 is within 2
            if event1.start > event2.start and event1.stop < event2.stop:
                return True
            if event1.stop > event2.start and event1.stop < event2.stop:
                return True

    return False

# allocate a y position for each gate
total_y = 0
for thread in threads:
    thread.total_y = total_y

    y = 1
    gate_positions = {}
    for event in thread.events:
        if event.work or event.wait:
            gate_positions[event.gate_name] = 0
        elif not event.gate_name in gate_positions:
            no_overlap = False
            for gate_name in gate_positions:
                if not is_overlap(thread.events, gate_name, event.gate_name):
                    gate_positions[event.gate_name] = gate_positions[gate_name]
                    no_overlap = True
                    break

            if not no_overlap:
                gate_positions[event.gate_name] = y
                y += 1

        event.y = gate_positions[event.gate_name]
        event.total_y = total_y + event.y

    total_y += y

PIXELS_PER_SECOND = 1000
PIXELS_PER_GATE = 20
LEFT_BORDER = 320
BAR_HEIGHT = 5
WIDTH = int(LEFT_BORDER + last_time * PIXELS_PER_SECOND)
HEIGHT = int((total_y + 1) * PIXELS_PER_GATE)

surface = cairo.ImageSurface(cairo.FORMAT_ARGB32, WIDTH + 50, HEIGHT)
ctx = cairo.Context(surface)
ctx.select_font_face('Sans')
ctx.set_font_size(15)

def draw_event(ctx, event):
    left = event.start * PIXELS_PER_SECOND + LEFT_BORDER
    top = event.total_y * PIXELS_PER_GATE 
    width = (event.stop - event.start) * PIXELS_PER_SECOND - 1
    height = BAR_HEIGHT

    ctx.rectangle(left, top, width, height)

    if event.wait:
        ctx.set_source_rgb(0.9, 0.1, 0.1)
    elif event.work:
        ctx.set_source_rgb(0.1, 0.9, 0.1)
    else:
        ctx.set_source_rgb(0.1, 0.1, 0.9)

    ctx.fill()

    if not event.wait and not event.work:
        xbearing, ybearing, twidth, theight, xadvance, yadvance = \
                ctx.text_extents(event.gate_name)

        ctx.move_to(left + width / 2 - twidth / 2, top + theight)
        ctx.set_source_rgb(1.00, 0.83, 0.00)
        ctx.show_text(event.gate_name)
        #ctx.stroke()

for thread in threads:
    xbearing, ybearing, twidth, theight, xadvance, yadvance = \
            ctx.text_extents(thread.thread_name)
    ctx.move_to(0, theight + thread.total_y * PIXELS_PER_GATE)
    ctx.set_source_rgb(1.00, 1.00, 1.00)
    ctx.show_text(thread.thread_name)
    #ctx.stroke()

    for event in thread.events:
        draw_event(ctx, event)

output_filename = "example.png"
print 'writing to', output_filename
surface.write_to_png(output_filename) 
